graph/src/components/ethereum/types.rs

```diff
@@ -1,12 +1,96 @@
 use ethabi::LogParam;
 use serde::{Deserialize, Serialize};
+use std::cmp::Ordering;
 use web3::types::*;
 
+pub type ThinEthereumBlock = Block<Transaction>;
+
+pub trait ThinEthereumBlockExt {
+    fn number(&self) -> u64;
+    fn transaction_for_log(&self, log: &Log) -> Option<Transaction>;
+    fn transaction_for_call(&self, call: &EthereumCall) -> Option<Transaction>;
+    fn parent_ptr(&self) -> EthereumBlockPointer;
+}
+
+impl ThinEthereumBlockExt for ThinEthereumBlock {
+    fn number(&self) -> u64 {
+        self.number.unwrap().as_u64()
+    }
+
+    fn transaction_for_log(&self, log: &Log) -> Option<Transaction> {
+        log.transaction_hash
+            .and_then(|hash| self.transactions.iter().find(|tx| tx.hash == hash))
+            .cloned()
+    }
+
+    fn transaction_for_call(&self, call: &EthereumCall) -> Option<Transaction> {
+        call.transaction_hash
+            .and_then(|hash| self.transactions.iter().find(|tx| tx.hash == hash))
+            .cloned()
+    }
+
+    fn parent_ptr(&self) -> EthereumBlockPointer {
+        EthereumBlockPointer {
+            hash: self.parent_hash,
+            number: self.number() - 1,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum BlockFinality {
+    /// If a block is final, we only need the header and the triggers.
+    Final(ThinEthereumBlock),
+
+    // If a block may still be reorged, we need to work with more local data.
```

Comment:
By `local data` you mean that we may still have to load the full block from Ethereum when processing the block, rather than already having everything in the database?

I still can't quite follow how `Final` vs. `NonFinal` are used throughout the code base and what makes them different. I think some more documentation would be good here.

One small request: Could we rename `ThinEthereumBlock` to `LightEthereumBlock`? Even though used in a different context, the light vs. full nomenclature is already established in the Ethereum space and feels more natural here. Even outside this context, "light" is probably more commonly used than "thin" (apart from when talking about wrappers and layers).

Reply:
That is a better name, I've renamed it.

I added a comment above `BlockFinality` to clarify. Actually I don't expect this enum to be around for long, for #902 I imagine we'll want a single trigger scanning function that can validate the block hashes and signal a revert if necessary. But to avoid this change becoming bigger than it already is, I introduced this enum which does the job for our current need of re-scanning only one block.

Reply:
Aaaah, now I get it. Thanks for the extra comment!